### PR TITLE
SLLS-86 Get remote project names from keys

### PR DESCRIPTION
**Pull request by jblievremont**

### CODE REVIEW

File: src/main/java/org/sonarsource/sonarlint/ls/connected/ProjectBindingManager.java [hunk 1]
```diff
@@ -621,4 +622,17 @@ public void run() {
       syncStorage();
     }
   }
+
+  public Map<String, String> getRemoteProjects(String connectionId) {
+    var serverConfiguration = getServerConfigurationFor(connectionId);
```
Comment (sophio-japharidze-sonarsource) [inline src/main/java/org/sonarsource/sonarlint/ls/connected/ProjectBindingManager.java:627]: Should this maybe be surrounded with a `try..catch` ? And also return an empty map in the catch block. Because I've seen a TimeOut exception thrown from the getConfiguration method, I think `could not get settings in time` or similar...

File: src/test/java/org/sonarsource/sonarlint/ls/mediumtests/ConnectedModeMediumTests.java [hunk 2]
```diff
@@ -165,4 +178,18 @@ void analysisConnected_matching_server_issues() throws Exception {
...
+    // Trigger a binding update to fetch projects in connected mode storage
+    ExecuteCommandParams updateBindings = new ExecuteCommandParams();
+    updateBindings.setCommand("SonarLint.UpdateAllBindings");
+    lsProxy.getWorkspaceService().executeCommand(updateBindings);
+
+    var params = new GetRemoteProjectsNamesParams(CONNECTION_ID, List.of(PROJECT_KEY1, "unknown"));
+
+    // Update of storage is asynchronous, eventually we get the right data in the storage :)
+    awaitUntilAsserted(() -> assertThat(lsProxy.getRemoteProjectNames(params).get()).containsExactly(Map.entry(PROJECT_KEY1, PROJECT_NAME1)));
+  }
```
Comment (sophio-japharidze-sonarsource) [inline src/test/java/org/sonarsource/sonarlint/ls/mediumtests/ConnectedModeMediumTests.java:194]: Do we need tests for when `connectionId` is undefined? Or when the `projectKeys` map is empty? Or when there is no match between the projects response and the projectKeys

**Review by sophio-japharidze-sonarsource — COMMENTED** [review]
Looks good with suggestions

**Review by kirill-knize-sonarsource — APPROVED** [review]
LGTM!

File: src/main/java/org/sonarsource/sonarlint/ls/connected/ProjectBindingManager.java (hunk 1 above)
Comment (jblievremont) [reply src/main/java/org/sonarsource/sonarlint/ls/connected/ProjectBindingManager.java:627]: I believe this is only supposed to happen if there was an error during the LS startup. I suggest we merge this and see if it happens frequently in dogfood.

**Review by sophio-japharidze-sonarsource — APPROVED** [review]

File: src/test/java/org/sonarsource/sonarlint/ls/mediumtests/ConnectedModeMediumTests.java (hunk 2 above)
Comment (jblievremont) [reply src/test/java/org/sonarsource/sonarlint/ls/mediumtests/ConnectedModeMediumTests.java:194]: The `connectionId == null` is indeed a corner case, thanks for noticing!

**Comment by jblievremont** [comment]
I had a look at how the download of projects is done in sonarlint-intellij, and I found that the previous implementation had an issue: the local cache of server projects is not necessarily available, especially the first time a user tries to bind to a server.

This new implementation explicitly asks the engine to download the list of projects and explicitly fails the request in case anything wrong happens (no more silent failure with an empty result).

File: src/main/java/org/sonarsource/sonarlint/ls/SonarLintLanguageServer.java [hunk 3]
```diff
@@ -448,4 +452,21 @@ public void onTokenUpdate() {
...
+      return CompletableFuture.completedFuture(
+        bindingManager.getRemoteProjects(params.getConnectionId())
+          .entrySet()
+          .stream()
+          .filter(e -> params.getProjectKeys().contains(e.getKey()))
+          .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))
+      );
+    } catch (IllegalStateException | IllegalArgumentException failed) {
+      var responseError = new ResponseError(ResponseErrorCode.InternalError, "Could not get remote project names", failed);
+      return CompletableFuture.failedFuture(new ResponseErrorException(responseError));
```
Comment (sophio-japharidze-sonarsource) [inline src/main/java/org/sonarsource/sonarlint/ls/SonarLintLanguageServer.java:468]: 👍🏻

**Review by sophio-japharidze-sonarsource — APPROVED** [review]
LGTM
